// @carbon/react/es/components/Menu/MenuContext.js
/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { objectSpread2 as _objectSpread2, toConsumableArray as _toConsumableArray } from '../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';

var menuDefaultState = {
  isRoot: true,
  hasIcons: false,
  size: null,
  items: [],
  requestCloseRoot: function requestCloseRoot() {}
};

function menuReducer(state, action) {
  switch (action.type) {
    case 'enableIcons':
      return _objectSpread2(_objectSpread2({}, state), {}, {
        hasIcons: true
      });

    case 'registerItem':
      return _objectSpread2(_objectSpread2({}, state), {}, {
        items: [].concat(_toConsumableArray(state.items), [action.payload]).filter(function (item) {
          return item.ref.current !== null;
        })
      });
  }
}

var MenuContext = /*#__PURE__*/React__default.createContext({
  state: menuDefaultState,
  // 'dispatch' is populated by the root menu
  dispatch: function dispatch() {}
});

export { MenuContext, menuReducer };
